File: android/app/src/main/java/com/clmix/SessionStore.kt
package com.clmix

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

// Holds only the server-issued session token (see RemoteServer._sessions in
// services/remote_server.py) - never the account password, which this app
// never writes to disk at all. Backed by EncryptedSharedPreferences (a
// Keystore-derived key encrypts the file at rest) since the token is a
// bearer credential: anyone who reads it can act as this user until the
// desktop app restarts or the user explicitly logs out.
object SessionStore {
    private const val TAG = "SessionStore"
    private const val FILE_NAME = "session"
    private const val KEY_TOKEN = "token"

    private fun prefs(context: Context): SharedPreferences {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        return EncryptedSharedPreferences.create(
            context,
            FILE_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    // The Keystore-backed cipher underneath EncryptedSharedPreferences is
    // known to throw on some OEM Keystore implementations (observed on
    // several Samsung/foldable builds) - most often after the device's
    // lock-screen credential changes and invalidates the existing key.
    // Session persistence is a convenience, not something worth crashing
    // the whole app over (this runs unconditionally on every launch, from
    // ConnectActivity.onCreate), so any failure here is treated the same
    // as "no saved session" rather than propagating, and the on-disk file
    // is wiped so a corrupt one doesn't keep failing on every future
    // launch.
    fun getToken(context: Context): String? = try {
        prefs(context).getString(KEY_TOKEN, null)
    } catch (e: Exception) {
        Log.w(TAG, "Failed to read session token, clearing stored session", e)
        context.deleteSharedPreferences(FILE_NAME)
        null
    }

    fun saveToken(context: Context, token: String) {
        try {
            prefs(context).edit().putString(KEY_TOKEN, token).apply()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to save session token", e)
        }
    }

    fun clear(context: Context) {
        try {
            prefs(context).edit().remove(KEY_TOKEN).apply()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to clear session token", e)
            context.deleteSharedPreferences(FILE_NAME)
        }
    }
}

File: ios/CLMix/SessionStore.swift
import Foundation
import Security

/// Holds only the server-issued session token (see RemoteServer._sessions
/// in services/remote_server.py) - never the account password, which this
/// app never writes to disk at all. Mirrors Android's SessionStore.kt,
/// with the Keychain standing in for EncryptedSharedPreferences: the token
/// is a bearer credential, so anyone who can read it can act as this user
/// until it expires (SESSION_TTL_SECONDS), the desktop app restarts, or
/// the user explicitly logs out.
///
/// `kSecAttrAccessibleAfterFirstUnlock` rather than `WhenUnlocked` so a
/// resume can still read the token when the app is relaunched into the
/// background (e.g. the reconnect path) on a device that hasn't been
/// unlocked since - it still keeps the token unreadable while the device
/// has never been unlocked after a boot.
enum SessionStore {
    private static let service = "com.clmix.session"
    private static let account = "token"

    private static var baseQuery: [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
    }

    static func token() -> String? {
        var query = baseQuery
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        guard status == errSecSuccess,
              let data = item as? Data,
              let token = String(data: data, encoding: .utf8),
              !token.isEmpty else {
            return nil
        }

        return token
    }

    static func save(_ token: String) {
        guard let data = token.data(using: .utf8) else { return }

        // Delete-then-add rather than SecItemUpdate: this runs on every
        // login (fresh or resumed), and an add over an existing item
        // fails with errSecDuplicateItem, so the delete makes the write
        // idempotent without a separate "does it exist" round trip.
        SecItemDelete(baseQuery as CFDictionary)

        var attributes = baseQuery
        attributes[kSecValueData as String] = data
        attributes[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock

        SecItemAdd(attributes as CFDictionary, nil)
    }

    static func clear() {
        SecItemDelete(baseQuery as CFDictionary)
    }
}
